cloud-controller-manager/do/firewall_controller.go

```diff
@@ -372,3 +493,38 @@ func (fc *firewallCache) updateCache(currentFirewall *godo.Firewall) {
 	defer fc.mu.Unlock()
 	fc.firewall = currentFirewall
 }
+
+func (fc *FirewallController) observeReconcileDuration(ctx context.Context, origin string) error {
+	var labels prometheus.Labels
+	switch {
+	case origin == originEventAdd:
+		labels = prometheus.Labels{"reconcile_type": originEventAdd}
+	case origin == originEventUpdate:
+		labels = prometheus.Labels{"reconcile_type": originEventUpdate}
+	case origin == originEventDelete:
+		labels = prometheus.Labels{"reconcile_type": originEventDelete}
+	case origin == originFirewallLoop:
+		labels = prometheus.Labels{"reconcile_type": originFirewallLoop}
+	default:
+		labels = prometheus.Labels{"reconcile_type": "none"}
+	}
+	t := prometheus.NewTimer(fc.fwManager.metrics.reconcileDuration.With(labels))
+	defer t.ObserveDuration()
+
+	return fc.ensureReconciledFirewall(ctx)
+}
+
+func (fc *FirewallController) observeRunLoopDuration(ctx context.Context, stopCh <-chan struct{}, fwReconcileFreq time.Duration) error {
+	labels := prometheus.Labels{"success": "false"}
+	t := prometheus.NewTimer(fc.fwManager.metrics.runLoopDuration.With(labels))
+	defer t.ObserveDuration()
+
+	err := fc.actualRun(stopCh, fwReconcileFreq)
+	if err == nil {
+		labels = prometheus.Labels{"success": "true"}
```

Comment:
Thank you for the example, that makes sense. Setting the variable again doesn't change `labels` before ObserveDuration is run, but because map is a reference type, changing that will have the desired effect.